File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SecondRound/Chapter12/rob_213.java
package algo.tzashinorpu.SecondRound.Chapter12;

public class rob_213 {
    public int rob(int[] nums) {
        int length = nums.length;
        if (length == 0) {
            return 0;
        }
        if (length == 1) {
            return nums[0];
        }
        int[] nums1 = new int[length - 1];
        int[] nums2 = new int[length - 1];
        System.arraycopy(nums, 0, nums1, 0, length - 1);
        System.arraycopy(nums, 1, nums2, 0, length - 1);
        return Math.max(robBottomUp(nums1), robBottomUp(nums2));
    }

    private int robBottomUp(int[] nums) {
        int prev = 0, curr = 0, temp = 0;
        for (int num : nums) {
            temp = curr;
            curr = Math.max(prev + num, curr);
            prev = temp;
        }
        return curr;
    }
}

File: JavascriptAlgorithm/DpPlan/I/rob_213.js
var rob = function (nums) {
  if (nums.length == 1) {
    return nums[0];
  }
  let nums1 = nums.slice(0, -1);
  let nums2 = nums.slice(1);
  function getMax(nums) {
    if (nums.length == 1) {
      return nums[0];
    } else {
      nums[1] = Math.max(nums[0], nums[1]);
    }
    for (let i = 2; i < nums.length; i++) {
      nums[i] = Math.max(nums[i - 1], nums[i - 2] + nums[i]);
    }
    return nums[nums.length - 1];
  }
  return Math.max(getMax(nums1), getMax(nums2));
};

export { rob };
